### ISSUE TITLE
Performance oddities

### ISSUE THREAD

**Issue by codeweaverx** [verement/decimal-arithmetic#2]
I have been experimenting with your library as a result of Pi day, and so a friend of mine and I were experimenting with Haskell and Python.  It turns out that it, too, has its Decimal library based on the General Decimal Arithmetic.

The first observation is in running two language versions of a gauss-legendre iterative solution, the python code outperforms the haskell code by a factor of 4.5 or so.  The target is to get to the first point at which the 'six nines' show up .

TheKnife:Python michaeljones$ python GaussLegendre.py 
...[snipped]
6086403441815981362977477130996051870721134999999 and so on!
0.40107011795 seconds

TheKnife:Code michaeljones$ ./GaussLegendre 
...[snipped]
6086403441815981362977477130996051870721134999999
Elapsed time: 1.825905s

These are pretty typical runs, though obviously not rigorously averaged over a lot of trials.

Granted, the Haskell program run above as a compiled executable was using ExtendedDecimal P1000 as its fixed decimal type.  But tightening that to P769, the minimum required to get the same output, I get something stranger:

TheKnife:Code michaeljones$ ./GaussLegendre 
...[snipped]
6086403441815981362977477130996051870721134999999
Elapsed time: 5.07026s

In other words, tightening the number of significant digits worsened the performance by a further factor of around 2.5.  This seems quite odd to me, but knowing why might mean I can mitigate it by choosing a possibly larger number of digits than I need because of implementation details.

Below, I'll provide the two programs involved.

I was running on OSX Sierra, GHC version 8.0.2, and Python 2.7.10.

[GaussLegendre.hs.zip](https://github.com/verement/decimal-arithmetic/files/1826193/GaussLegendre.hs.zip)
[GaussLegendre.py.zip](https://github.com/verement/decimal-arithmetic/files/1826199/GaussLegendre.py.zip)

**Comment by verement** [comment]
This line seems suspect:

```haskell
type P769 = PTimes2 P768
```

I think this is really P1536, which might explain the discrepancy.

**Comment by codeweaverx** [comment]
Quite so.  Good catch.  I stared at that forever and couldn't see the forest for the trees apparently.

That discrepancy is resolved.  I'm now at

TheKnife:Code michaeljones$ ./GaussLegendre 
Gauss-Legendre needs 10 to give six nines, Gauss-legendre, 10.  n == 10
...[snipped]
6086403441815981362977477130996051870721134999999
Elapsed time: 0.991275s

which is a considerable improvement.

The gap's been narrowed nicely.  

Thanks for the patience in spotting my foolish error :)
